src/token-group/internal.tsx

```diff
@@ -8,28 +8,51 @@ import checkControlled from '../internal/hooks/check-controlled';
 import { InternalBaseComponentProps } from '../internal/hooks/use-base-component';
 
 import { TokenGroupProps } from './interfaces';
-
 import { SomeRequired } from '../internal/types';
-import GenericTokenGroup from './generic-token-group';
+import { getBaseProps } from '../internal/base-component';
+import clsx from 'clsx';
+import styles from './styles.css.js';
+import TokenList from '../internal/components/token-list';
+import { Token } from './token';
 
 type InternalTokenGroupProps = SomeRequired<TokenGroupProps, 'items' | 'alignment'> & InternalBaseComponentProps;
 
-export default function InternalTokenGroup({ items, onDismiss, __internalRootRef, ...props }: InternalTokenGroupProps) {
+export default function InternalTokenGroup({
+  alignment,
+  items,
+  onDismiss,
+  limit,
+  i18nStrings,
+  __internalRootRef,
+  ...props
+}: InternalTokenGroupProps) {
   checkControlled('TokenGroup', 'items', items, 'onDismiss', onDismiss);
 
+  const baseProps = getBaseProps(props);
+  const hasItems = items.length > 0;
+
   return (
-    <GenericTokenGroup
+    <div
+      {...baseProps}
+      className={clsx(baseProps.className, styles.root, hasItems && styles['has-items'])}
       ref={__internalRootRef}
-      {...props}
-      items={items}
-      renderItem={item => <Option option={item} />}
-      getItemAttributes={(item, itemIndex) => ({
-        disabled: item.disabled,
-        dismiss: {
-          label: item.dismissLabel,
-          onDismiss: () => fireNonCancelableEvent(onDismiss, { itemIndex }),
-        },
-      })}
-    />
+    >
+      <TokenList
+        alignment={alignment}
+        items={items}
+        limit={limit}
+        renderItem={(item, itemIndex) => (
+          <Token
+            key={itemIndex}
```

Comment:
Currently the `key` property is implicitly required because the items are passed directly to the space-between. That is going to change after https://github.com/cloudscape-design/components/pull/876 is merged. Then, each item will be wrapped with `li` in the token-list so the keys will no longer be needed.